### DISCUSSION TITLE
Boost Controls Release of Deals to Sealing Subsystem based on Sealing Subsystem Load

### DISCUSSION THREAD

**Discussion by scaseye** [filecoin-project/boost#817] [Ideas]
Would be really great to be able to control the release of deals to the publish queue based on what is currently running on the sealing system.  I do this on offline deals via scripts.  If I have too many PC1s or GETS happening I do not import the next deal.

Would give an SP the ability to better control how many sectors are running at the same time in their stack.  Prevent overloads. 

Prior to putting a deal into the Boost Publish Queue
![image](https://user-images.githubusercontent.com/69066370/191095274-27c3e6dc-0d5d-47bd-a35b-aaee6fac7421.png)

Check Load of Stack.  If too many sectors in a stage, wait.  or if there is a Publish message still pending wait.

**Comment by LexLuthr** [comment]
Hey @scaseye, we have been working on this behind the scenes. This required some joint effort from lotus as well as boost team.

Magik has created https://github.com/filecoin-project/lotus/issues/8923 to expose the sealing pipeline stats. Once this is release, boost team will look into how to utilize them in the most efficient way to onboard deals.

**Comment by scaseye** [comment]
outstanding!

**Comment by scaseye** [comment]
@LexLuthr I can see all the pieces in the GraphQL API right now!

`query {
  sealingpipeline{
    Workers{
      Stage
    }
  }
  mpool(local:true){
    From
    Method
  }
}`

@dirkmc What would it take to have Boost NOT send a deal to the publish queue based the count of Workers.Stage, and if there is already a message in mpool PublishDeals?

new settings

- MAXSEALJOBS
- SEALJOBTYPESTOCOUNT

example:
MAXSEALJOBS = 50
SEALJOBTYPESTOCOUNT = "PC1\|RU\|GET\|AP\|FRU"

**Comment by scaseye** [comment]
`{
  "data": {
    "sealingpipeline": {
      "Workers": [
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "GET"
        },
        {
          "Stage": "AP"
        },
        {
          "Stage": "AP"
        },
        {
          "Stage": "AP"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "PC1"
        },
        {
          "Stage": "GET"
        },
        {
          "Stage": "PC2"
        }
      ]
    },
    "mpool": [
      {
        "From": "f3r7hmxkfa27uu74hgwuaqklcreipv6gs7pe...",
        "Method": "PublishDeal"
      }
    ]
  }
}
`
